api/server/router/volume/volume_routes.go

```diff
@@ -45,25 +117,101 @@ func (v *volumeRouter) postVolumesCreate(ctx context.Context, w http.ResponseWri
 	}
 
 	var req volume.CreateOptions
+	if err := httputils.CheckForJSON(r); err != nil {
+		return err
+	}
+
+	version := httputils.VersionFromContext(ctx)
+
 	if err := httputils.ReadJSON(r, &req); err != nil {
 		return err
 	}
 
-	vol, err := v.backend.Create(ctx, req.Name, req.Driver, opts.WithCreateOptions(req.DriverOpts), opts.WithCreateLabels(req.Labels))
+	var (
+		volume *volume.Volume
+		err    error
+	)
+
+	// if the ClusterVolumeSpec is filled in, then this is a cluster volume
+	// and is created through the swarm cluster volume backend.
+	//
+	// re: volume name duplication
+	//
+	// As it happens, there is no good way to prevent duplication of a volume
+	// name between local and cluster volumes. This is because Swarm volumes
+	// can be created from any manager node, bypassing most of the protections
+	// we could put into the engine side.
+	//
+	// Instead, we will allow creating a volume with a duplicate name, which
+	// should not break anything.
+	if req.ClusterVolumeSpec != nil && versions.GreaterThanOrEqualTo(version, clusterVolumesVersion) {
+		logrus.Debug("using cluster volume")
+		volume, err = v.cluster.CreateVolume(req)
+	} else {
+		logrus.Debug("using regular volume")
+		volume, err = v.backend.Create(ctx, req.Name, req.Driver, opts.WithCreateOptions(req.DriverOpts), opts.WithCreateLabels(req.Labels))
+	}
+
+	if err != nil {
+		return err
+	}
+	return httputils.WriteJSON(w, http.StatusCreated, volume)
+}
+
+func (v *volumeRouter) postVolumesUpdate(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
```

Comment:
Same here; we could use an early return (perhaps we should use the generic error though (see [`func (c *Cluster) errNoManager(st nodeState) error {`](https://github.com/moby/moby/blob/647aede6ad42d0f4cf08d675f59bb8013a3fb56b/daemon/cluster/cluster.go#L352-L354));

```suggestion
func (v *volumeRouter) postVolumesUpdate(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
	if !v.cluster.IsManager() {
		return errdefs.Unavailable(errors.New("volume update only valid for cluster volumes, but swarm unavailable"))
	}
```

```go
func (v *volumeRouter) postVolumesUpdate(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
	if !v.cluster.IsManager() {
		return errdefs.Unavailable(errors.New("volume update only valid for cluster volumes, but swarm unavailable"))
	}
	if err := httputils.ParseForm(r); err != nil {
		return err
	}
	if err := httputils.CheckForJSON(r); err != nil {
		return err
	}

	rawVersion := r.URL.Query().Get("version")
	version, err := strconv.ParseUint(rawVersion, 10, 64)
	if err != nil {
		err = fmt.Errorf("invalid swarm object version '%s': %v", rawVersion, err)
		return errdefs.InvalidParameter(err)
	}

	var req volume.VolumeUpdateBody
	if err := httputils.ReadJSON(r, &req); err != nil {
		return err
	}

	return v.cluster.UpdateVolume(vars["name"], version, req)
}
```